### ISSUE TITLE
CUSTOM_CHAR of type STRING leads to an error

### ISSUE THREAD

**Issue by planetk** [HomeSpan/HomeSpan#198]
Hi,

first let me thank you for your great lib. I'm using release 1.4.2  Unfortunatly im experiencing an issue when I try to define a custom characteristic.

as soon as I add a line

CUSTOM_CHAR(LightMode, AAAAAAAA-AAAA-AAAA-AAAA-AAAAAAAAAAAA, PR, STRING, NULL, NULL, NULL, true); 
or
CUSTOM_CHAR(LightMode, AAAAAAAA-AAAA-AAAA-AAAA-AAAAAAAAAAAA, PR, STRING, "ANY_VALUE", NULL, NULL, true); 

I get compilation errors:

<...>/Arduino/libraries/HomeSpan/src/HomeSpan.h:335:17: error: cast from 'char*' to 'uint8_t' {aka 'unsigned char'} loses precision [-fpermissive]
         u.UINT8=(uint8_t)val;
                 ^~~~~~~~~~~~
<...>/Arduino/libraries/HomeSpan/src/HomeSpan.h:338:18: error: cast from 'char*' to 'uint16_t' {aka 'short unsigned int'} loses precision [-fpermissive]
         u.UINT16=(uint16_t)val;
                  ^~~~~~~~~~~~~
<...>/Arduino/libraries/HomeSpan/src/HomeSpan.h:347:17: error: invalid cast from type 'char*' to type 'double'
         u.FLOAT=(double)val;
                 ^~~~~~~~~~~

Is there a way to work around this?

Thanks for any help.

Regards
Stefan

**Comment by HomeSpan** [comment]
Thanks for catching this.  I am able to reproduce and will debug.

**Comment by HomeSpan** [comment]
I've identified and the issue and will fix in the next patch release.  In the interim, a simple workaround is override the definition of STRING_t (which is declared as a typedef to "char *" instead of "const char *" for other reasons) by inserting the following line prior to calling CUSTOM_CHAR:

```C++
#define STRING_t  const char *              // WORK-AROUND

CUSTOM_CHAR(LightMode, AAAAAAAA-AAAA-AAAA-AAAA-AAAAAAAAAAAA, PR, STRING, "ANY_VALUE", NULL, NULL, true);
```

**Comment by HomeSpan** [comment]
This has been fixed - a new macro, `CUSTOM_CHAR_STRING()` has been added to specifically handle String Characteristics.  I will add to next patch release, but it is available now in the dev branch it you'd like to try.  See the API in the dev branch for details.

**Comment by planetk** [comment]
Wow, thank you very much for the quick fix! The solution sounds good to me, although I have to admit that I just had a quick glance at it yet. I'll give it a try within the next few days.

What comes into my mind is, if it would be useful to add a max length to the String Characteristics, as e.g. the Apple Homekit Accessory Simulator suggests. Anyhow, length seems to default to 32 which is pretty much fine for me.

**Comment by HomeSpan** [comment]
HomeSpan itself does not impose any limits to a String Characteristic since memory is allocated as needed.  However HAP-R2 does impose a limit - according to the HAP doc, the default *maxLen* is 64 characters, though this can be changed to allow up to a *maxLen* of 256 characters.   It would be a relatively straightforward change to add a method to change the *maxLen* parameter, but it sounds like 64 will be more than sufficient for your needs.  May I ask how you are using a custom String Characteristic?  Does this work with the Eve HomeKit App or another Home App?

**Comment by planetk** [comment]
Hi,
Thanks for the clarification. Currently I habe two different projects  for the custom characteristics.

First I did some reverse engineering on the elgato eve custom characteristics for a previous homebridge project. I‘d like to port these characteristics to a stand alone esp32 solution, so that the Eve App will be able to pick up the historic data.

The specific project I currently need the String characteristics for is comparable to the nanoleaf HomeKit integration. When storing scenes with nanoleaf devices, also the current animation is stored with a scene. I‘m not totally sure, if this works just with the Home App, or if I have to provide a small Swift companion app.

I‘ll let you know in the discussions section, as soon as I got some progress on one of the projects.

Thanks again and kind regards!
